src/OrleansRuntime/Services/GrainServiceClient.cs

```diff
@@ -0,0 +1,67 @@
+﻿using Orleans.CodeGeneration;
+using Orleans.Runtime.ConsistentRing;
+using Orleans.Services;
+
+namespace Orleans.Runtime.Services
+{
+    /// <summary>
+    /// Proxies requests to the appropriate GrainService based on the appropriate Ring partitioning strategy.
+    /// </summary>
+    /// <typeparam name="TGrainService"></typeparam>
+    public abstract class GrainServiceClient<TGrainService> : IGrainServiceClient<TGrainService> where TGrainService : IGrainService
+    {
+        private readonly IConsistentRingProvider ringProvider;
+        private readonly int grainTypeCode;
+
+        /// <summary>
+        /// Currently we only support a single GrainService per Silo, when multiple are supported we will request the number of GrainServices to partition per silo here.
+        /// </summary>
+        protected GrainServiceClient()
+        {
+            ringProvider = (IConsistentRingProvider) Silo.CurrentSilo.Services.GetService(typeof(IConsistentRingProvider));
+            
+            // GrainInterfaceMap only holds IGrain types, not ISystemTarget types, so resolved via Orleans.CodeGeneration.
+            // Resolve this before merge.
+            grainTypeCode = GrainInterfaceUtils.GetGrainClassTypeCode(typeof(TGrainService));
+        }
+
+        /// <summary>
+        /// Resolves the correct GrainService responsible for actioning the request based on the CallingGrainReference
+        /// </summary>
+        protected TGrainService GrainService
+        {
+            get
+            {
+                // Currently we only support a single GrainService per Silo, hence we pass Zero here instead of partitioning.
+                var grainId = GrainId.GetGrainServiceGrainId(0, grainTypeCode);
+                
+                var destination = MapGrainReferenceToSiloRing(CallingGrainReference);
+                var grainService = RuntimeClient.Current.InternalGrainFactory.GetSystemTarget<TGrainService>(grainId, destination);
+
+                return grainService;
+            }
+        }
+        
+        /// <summary>
+        /// Resolves the Grain Reference invoking this request.
+        /// </summary>
+        protected GrainReference CallingGrainReference
+        {
+            get
+            {
+                return RuntimeClient.Current.CurrentActivationData.GrainReference;
```

Comment:
I'm trying to remove statics like this so that we can enable non-static grain clients